File: sentimental-credit/credit.py
from cs50 import get_int


def main():
    cardnumber = str(get_int("Number: "))  # Works with digits as a string, digits are converted back to int when necessary
    len_cardnumber = len(cardnumber)
    initial_digits = int(cardnumber[0:2])
    initial_digit = int(cardnumber[0])
    checksum = f_checksum(cardnumber)

    if len_cardnumber == 15 and (initial_digits == 34 or initial_digits == 37) and checksum == 0:  # Conditions for AMEX.
        print("AMEX")
    elif len_cardnumber == 16 and (51 <= initial_digits <= 55) and checksum == 0:  # Conditions for MASTERCARD.
        print("MASTERCARD")
    elif (len_cardnumber == 13 or len_cardnumber == 16) and initial_digit == 4 and checksum == 0:  # Conditions for VISA.
        print("VISA")
    else:  # Else INVALID.
        print("INVALID")


def f_checksum(cardnumber):
    sum_of_products_of_even_digits = 0
    for i in range(len(cardnumber) - 2, -1, -2):  # Digits in even position starting from second to last
        product_of_digit = str(int(cardnumber[i]) * 2)
        for digit in product_of_digit:
            sum_of_products_of_even_digits += int(digit)

    sum_of_odd_digits = 0
    for i in range(len(cardnumber) - 1, -1, -2):  # Digits in odd digits startting from last
        sum_of_odd_digits += int(cardnumber[i])

    return int(str(sum_of_products_of_even_digits + sum_of_odd_digits)[-1])


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

long ask_cardnumber(void);
int lenth(long number);
int initial_digits(long cardnumber, int lenthcard, int slice);
int check_sum(long cardnumber);


int main(void)
{
    long cardnumber = ask_cardnumber(); //asks users for card
    int lenthcard = lenth(cardnumber); //returns the number of digits in card
    int initialDigits = initial_digits(cardnumber, lenthcard, 2); // returns 2 initial digits
    int initialDigit = initial_digits(cardnumber, lenthcard, 1); // returns 1 initial digit
    int checksum = check_sum(cardnumber); //checksum of card

    if (lenthcard == 15 && (initialDigits == 34 || initialDigits == 37) && checksum == 0) //conditions for amex
    {
        printf("AMEX\n");
    }
    else if (lenthcard == 16 && initialDigits >= 51 && initialDigits <= 55 && checksum == 0) //conditions for master
    {
        printf("MASTERCARD\n");
    }
    else if ((lenthcard == 13 || lenthcard == 16) && initialDigit == 4 && checksum == 0) //conditions for visa
    {
        printf("VISA\n");
    }
    else //else invalid
    {
        printf("INVALID\n");
    }
}

long ask_cardnumber(void)
{
    return get_long("Card Number: ");
}

int lenth(long cardnumber)
{
    int lenthcard;
    for (lenthcard = 0; cardnumber != 0; cardnumber /= 10)
    {
        lenthcard++;
    }
    return lenthcard;
}

int initial_digits(long cardnumber, int lenthcard, int slice)
{
    return (int)(cardnumber / pow(10, lenthcard - slice));
}

int check_sum(long cardnumber)
{
    int checksum = 0;
    for (int digitPosition = 1; cardnumber != 0; digitPosition++) //digitspositions until number lenth
    {
        int digit = cardnumber % 10; //digit value
        cardnumber /= 10;

        if (digitPosition % 2 == 0) //if position is even
        {
            int digitresult = digit * 2; //multiplie by 2
            for (int digitP = 1; digitresult != 0; digitresult /= 10, digitP++) //taks each digit in resul and add to checksum
            {
                checksum = checksum + digitresult % 10;
            }
        }
        else //just add digit to sum
        {
            checksum = checksum + digit;
        }
    }
    return checksum % 10; //returns 0 if ends in 0
}